### PR TITLE
Xml cnn

### PR DESCRIPTION
**Pull request by sian-chen**
1. Implement KimCNN, XMLCNN
2. Remove get_network(), use model_name to get the network instead
3. Allow setting weight initialization method
4. Add config to log
5. Allow fixed length (for dynamic max-pooling)
6. Add config files for EUR-Lex

### CODE REVIEW

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

File: model.py [hunk 1]
```diff
@@ -52,12 +52,15 @@ def __init__(self, config, word_dict=None, classes=None, ckpt=None):
...
-        self.network = get_network(config, embed_vecs).to(self.device)
+        self.network = getattr(networks, config.model_name)(config, embed_vecs).to(self.device)
         self.init_optimizer()
 
         if ckpt:
             self.network.load_state_dict(ckpt['state_dict'])
             self.optimizer.load_state_dict(ckpt['optimizer'])
+        else:
+            init_weights = networks.get_weight_init_func(config)
+            self.network.apply(init_weights)
```
[the remark this answers is not in the document]
Comment (Eleven1Liu) [reply model.py:63]: They said "Pass an initialization function to `torch.nn.Module.apply`. It will initialize the weights in the entire `nn.Module` recursively."
https://stackoverflow.com/questions/49433936/how-to-initialize-weights-in-pytorch

Is that the reason the linear layers start from different weights?

File: example_config/EUR-Lex/kim_cnn.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#filter_size: 4 # no filter sizes
+filter_sizes: [2, 4, 8]
+strides: [1, 1, 1]
+dropout: 0.2
+
+# pretrained vocab / embeddings
+embed_file: glove.6B.300d
+
+# others
+data_workers: 4
```
Comment (cjlin1) [inline example_config/EUR-Lex/kim_cnn.yml:33]: this is the default value. Do we now need this in all yml files?

File: networks/__init__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+from networks.caml import CAML
+from networks.kim_cnn import KimCNN
+from networks.xml_cnn import XMLCNN
+
+
+def get_weight_init_func(config):
+    def init_weights(m):
+        if isinstance(m, nn.Linear) or isinstance(m, nn.Conv2d):
+            getattr(nn.init, config.weight_init + '_')(m.weight)
+    return init_weights
```
Comment (cjlin1) [inline networks/__init__.py:12]: why don't we just call this fun get_init_weights()??

File: networks/__init__.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import torch.nn as nn
+
+from networks.caml import CAML
+from networks.kim_cnn import KimCNN
+from networks.xml_cnn import XMLCNN
+
+
+def get_weight_init_func(config):
+    def init_weights(m):
```
Comment (cjlin1) [inline networks/__init__.py:9]: we also have --weight-init. For consistency maybe we should always use "init_weights"?

File: model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cjlin1) [reply model.py:63]: I have the same concern.. Are we able to get same results for mimic-50 as before?

File: networks/xml_cnn.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class XMLCNN(BaseModel):
+    def __init__(self, config, embed_vecs):
+        super(XMLCNN, self).__init__(config, embed_vecs)
+        assert config.fixed_length is True
+
+        self.filter_sizes = config.filter_sizes
+        emb_dim = embed_vecs.shape[1]
+        num_filter_maps = config.num_filter_maps
```
Comment (cjlin1) [inline networks/xml_cnn.py:21]: "filter_maps"  may not be a good name.  I guess we can have
num_filters_per_size
and this should be put right after
filter_sizes
in yml file

File: networks/xml_cnn.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        self.convs = nn.ModuleList()
+        self.poolings = nn.ModuleList()
+
+        for filter_size, p, stride in zip(self.filter_sizes, d_max_pool_p, strides):
+            conv = nn.Conv2d(
+                in_channels=1,
+                out_channels=num_filter_maps,
+                kernel_size=(filter_size, emb_dim),
+                stride=(stride, emb_dim))
```
Comment (cjlin1) [inline networks/xml_cnn.py:33]: This can is indeed a Conv1d operation.. Not sure if we want to do 2d here?? Need some discussion I think

File: networks/kim_cnn.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        strides = config.strides
+
+        self.convs = nn.ModuleList()
+
+        for filter_size, stride in zip(self.filter_sizes, strides):
+            conv = nn.Conv2d(
+                in_channels=1,
+                out_channels=num_filter_maps,
+                kernel_size=(filter_size, emb_dim),
+                stride=(stride, emb_dim))
```
Comment (cjlin1) [inline networks/kim_cnn.py:24]: This can is indeed a Conv1d operation.. Not sure if we want to do 2d here?? Need some discussion I think

File: networks/kim_cnn.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+from networks.base import BaseModel
+
+
+class KimCNN(BaseModel):
+    def __init__(self, config, embed_vecs):
+        super(KimCNN, self).__init__(config, embed_vecs)
+
+        self.filter_sizes = config.filter_sizes
+        emb_dim = embed_vecs.shape[1]
+        num_filter_maps = config.num_filter_maps
```
Comment (cjlin1) [inline networks/kim_cnn.py:14]: "filter_maps" may not be a good name. I guess we can have
num_filters_per_size
and this should be put right after
filter_sizes
in yml file

File: networks/kim_cnn.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            h_sub = h_sub.view(h_sub.shape[0], h_sub.shape[1], -1) # (batch_size, num_filter, H)
+            h_sub = F.max_pool1d(h_sub, kernel_size=h_sub.size()[2]) # (batch_size, num_filter, 1)
+            h_sub = h_sub.view(h_sub.shape[0], -1) # (batch_size, num_filter)
+            h_list.append(h_sub)
+
+        if len(self.filter_sizes) > 1:
+            h = torch.cat(h_list, 1)
+        else:
+            h = h_list[0]
+        h = F.relu(h) # (batch_size, N * num_filter)
```
Comment (cjlin1) [inline networks/kim_cnn.py:47]: for mimic-50 we did tanh??

Comment (cjlin1) [reply networks/kim_cnn.py:47]: For "N * num_filter" we have no N
I think this is just conv_output_size?

File: model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sian-chen) [reply model.py:63]: Yes, we should set `weight_init: xavier_uniform` in `caml.yml` and remove `xavier_uniform_()` in `caml.py`.
However, since the order of initialization is changed, we may not able to get the same results for MIMIC-50 as before.
It is possible to get the same results but the code would be pretty ugly...

File: networks/kim_cnn.py (hunk 7 above)
Comment (sian-chen) [reply networks/kim_cnn.py:24]: Though what we exactly do is 1D-convolution, we have to use Conv2D here because our input is 2D (doc_len, emb_size).
Conv1D can only handle 1D input (doc_len).

File: networks/__init__.py (hunk 3 above)
Comment (sian-chen) [reply networks/__init__.py:12]: The behavior of the function depends on `config`, so we need a function generator.

File: networks/kim_cnn.py (hunk 9 above)
Comment (sian-chen) [reply networks/kim_cnn.py:47]: Yes, it is conv_output_size = num_filter_size * num_filter_maps = 3 * 128.

File: networks/kim_cnn.py (hunk 7 above)
Comment (sian-chen) [reply networks/kim_cnn.py:24]: Oh sorry, you're correct. We can transpose the input to treat emb_size as the number of input channels then use Conv1D.

File: model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sian-chen) [reply model.py]: Solved.

File: example_config/EUR-Lex/kim_cnn.yml (hunk 2 above)
Comment (sian-chen) [reply example_config/EUR-Lex/kim_cnn.yml]: Solved.

File: networks/__init__.py (hunk 4 above)
Comment (sian-chen) [reply networks/__init__.py]: Use "init_weight" in all places.

File: networks/kim_cnn.py (hunk 9 above)
Comment (sian-chen) [reply networks/kim_cnn.py]: Change "N * num_filter" to "total_num_filter".

File: data_utils.py [hunk 10]
```diff
@@ -67,61 +69,65 @@ def tokenize(text):
...
 
 
 def load_or_build_text_dict(config, dataset):
     if config.vocab_file:
-        log.info(f'Load vocab from {config.vocab_file}')
+        logging.info(f'Load vocab from {config.vocab_file}')
         with open(config.vocab_file, 'r') as fp:
-            vocab_list = ['**PAD**'] + [vocab.strip() for vocab in fp.readlines()]
-        vocabs = Vocab(collections.Counter(vocab_list), specials=['<unk>'],
+            vocab_list = [PAD] + [vocab.strip() for vocab in fp.readlines()]
```
Comment (cjlin1) [inline data_utils.py:114]: I suggest we add a comment here:

# Due to the tentative need to reproduce CAML results, if vocab_file is used,
# PAD is put in the beginning instead of in the end

File: networks/__init__.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import torch.nn as nn
+
+from networks.caml import CAML
+from networks.kim_cnn import KimCNN
+from networks.xml_cnn import XMLCNN
+
+
+def get_init_weight_func(config):
+    def init_weight(m):
+        if isinstance(m, nn.Linear) or isinstance(m, nn.Conv2d):
```
Comment (cjlin1) [inline networks/__init__.py:10]: but now we do Conv1d?

File: networks/__init__.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+            getattr(nn.init, config.init_weight+ '_')(m.weight)
```
Comment (cjlin1) [inline networks/__init__.py:11]: If  the "if" statement fails (e.g., Conv1d above), should we issue an error/warning message?

File: networks/xml_cnn.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from networks.base import BaseModel
+
+
+def out_size(l_in, kernel_size, channels, padding=0, dilation=1, stride=1):
+    a = l_in + 2 * padding - dilation * (kernel_size - 1) - 1
+    b = int(a / stride)
```
Comment (cjlin1) [inline networks/xml_cnn.py:10]: since we have l_in, we can have l_out
and
l_out = math.floor(a/stride)+1
we can then return
l_out * channels
BTW I prefer floor over int for better readability
Also can add a comment to refer to
https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html
about the l_out calculation

File: networks/__init__.py (hunk 11 above)
Comment (Eleven1Liu) [reply networks/__init__.py:10]: It seems that Si-An has added `nn.Conv1d` here locally. @sian-chen

File: networks/xml_cnn.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+        for filter_size, p, stride in zip(self.filter_sizes, d_max_pool_p, strides):
+            conv = nn.Conv1d(
+                in_channels=emb_dim,
+                out_channels=num_filter_maps,
+                kernel_size=filter_size,
+                stride=stride)
+
+            # Dynamic Max-Pooling
+            conv_out_size = out_size(config.max_seq_length, filter_size, num_filter_maps, stride=stride)
```
Comment (cjlin1) [inline networks/xml_cnn.py:36]: we have conv_out_size and
conv_output_size below. Can just use one name (though for different things)
Maybe conv_output_size we have used that in kimcnn?

File: networks/xml_cnn.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    def forward(self, text):
+        h = self.embedding(text) # (batch_size, length, embed_dim)
+        h = self.embed_drop(h)
+        h = h.view(h.shape[0], h.shape[2], h.shape[1]) # (batch_size, embed_dim, length)
+
+        h_list = []
+        for i in range(len(self.filter_sizes)):
+            h_sub = self.convs[i](h) # (batch_size, num_filter, length)
+            h_sub = h_sub.view(h_sub.shape[0], 1, h_sub.shape[1] * h_sub.shape[2]) # (batch_size, 1, num_filter * length)
+            h_sub = self.poolings[i](h_sub) # (batch_size, 1, P)
```
Comment (cjlin1) [inline networks/xml_cnn.py:57]: we used p instead of P earlier?

File: networks/xml_cnn.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            h_sub = h_sub.view(h_sub.shape[0], -1) # (batch_size, P)
+            h_list.append(h_sub)
+
+        if len(self.filter_sizes) > 1:
+            h = torch.cat(h_list, 1)
+        else:
+            h = h_list[0]
+        h = self.activation(h) # (batch_size, N * num_filter)
```
Comment (cjlin1) [inline networks/xml_cnn.py:65]: what's N? Can we use the same description in line 46 of kim_cnn?

**Review by cjlin1 — APPROVED** [review]

File: networks/kim_cnn.py (hunk 7 above)
Comment (sian-chen) [reply networks/kim_cnn.py]: I found using Conv1d lead to significantly worse results.
Need further investigation...

File: data_utils.py (hunk 10 above)
Comment (sian-chen) [reply data_utils.py:114]: Why we need this comment?
For the other case (vocab_file is not used), PAD is also put in the beginning.

File: networks/__init__.py (hunk 11 above)
Comment (sian-chen) [reply networks/__init__.py:10]: Yes, I've added Con1d here.

File: networks/__init__.py (hunk 12 above)
Comment (sian-chen) [reply networks/__init__.py:11]: For the rest of the modules like BatchNorm, Dropout, etc, people usually use the default initialization.
So I think we don't need to raise an error/warning here.

File: networks/kim_cnn.py (hunk 7 above)
Comment (sian-chen) [reply networks/kim_cnn.py]: Solved. Should transpose the input for Conv1d.

File: networks/kim_cnn.py (hunk 8 above)
Comment (sian-chen) [reply networks/kim_cnn.py:14]: Solved.

File: networks/xml_cnn.py (hunk 13 above)
Comment (sian-chen) [reply networks/xml_cnn.py:10]: Solved.

File: networks/xml_cnn.py (hunk 14 above)
Comment (sian-chen) [reply networks/xml_cnn.py:36]: Solved.

File: networks/xml_cnn.py (hunk 15 above)
Comment (sian-chen) [reply networks/xml_cnn.py:57]: Solved.

File: networks/xml_cnn.py (hunk 16 above)
Comment (sian-chen) [reply networks/xml_cnn.py:65]: Solved.

**Comment by cjlin1** [comment]
On 2021-04-30 16:25, sianchen wrote:
Yes, you are right.. So please modify the comment to say the difference 
you mentioned

File: networks/xml_cnn.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    return l_out * channels
+
+
+class XMLCNN(BaseModel):
+    def __init__(self, config, embed_vecs):
+        super(XMLCNN, self).__init__(config, embed_vecs)
+        assert config.fixed_length is True
+
+        self.filter_sizes = config.filter_sizes
+        emb_dim = embed_vecs.shape[1]
```
Comment (cjlin1) [inline networks/xml_cnn.py:23]: 23 and 24 can be swapped so people can connect 22 and 23 together

File: networks/__init__.py (hunk 12 above)
Comment (cjlin1) [reply networks/__init__.py:11]: OK

File: data_utils.py (hunk 10 above)
Comment (cjlin1) [reply data_utils.py:114]: Yes, you are right.. So please modify the comment to say the difference you mentioned

File: networks/xml_cnn.py (hunk 17 above)
Comment (sian-chen) [reply networks/xml_cnn.py]: Solved.

File: data_utils.py (hunk 10 above)
Comment (sian-chen) [reply data_utils.py:114]: OK. Done.

**Review by Eleven1Liu — APPROVED** [review]
